src/EditorFeatures/CSharpTest/Completion/CompletionProviders/CompletionProviderOrderTests.cs

```diff
@@ -0,0 +1,141 @@
+{
+    [UseExportProvider]
+    public class CompletionProviderOrderTests
+    {
+        [Fact]
+        public void TestCompletionProviderOrder()
+        {
+            var completionProviderExports = TestExportProvider.ExportProviderWithCSharpAndVisualBasic.GetExports<CompletionProvider, CompletionProviderMetadata>();
+            var orderedCSharpCompletionProviders = ExtensionOrderer.Order(completionProviderExports.Where(export => export.Metadata.Language == LanguageNames.CSharp));
+
+            var actualOrder = orderedCSharpCompletionProviders.Select(x => x.Value.GetType()).ToArray();
+            var expectedOrder = new[]
+            {
+                // Marker for start of built-in completion providers
+                typeof(FirstBuiltInCompletionProvider),
+
+                // Built-in providers
+                typeof(AttributeNamedParameterCompletionProvider),
+                typeof(NamedParameterCompletionProvider),
+                typeof(KeywordCompletionProvider),
+                typeof(SpeculativeTCompletionProvider),
+                typeof(SymbolCompletionProvider),
+                typeof(ExplicitInterfaceMemberCompletionProvider),
+                typeof(ExplicitInterfaceTypeCompletionProvider),
+                typeof(ObjectCreationCompletionProvider),
+                typeof(ObjectInitializerCompletionProvider),
+                typeof(CSharpSuggestionModeCompletionProvider),
+                typeof(EnumAndCompletionListTagCompletionProvider),
+                typeof(CrefCompletionProvider),
+                typeof(SnippetCompletionProvider),
+                typeof(ExternAliasCompletionProvider),
+                typeof(OverrideCompletionProvider),
+                typeof(PartialMethodCompletionProvider),
+                typeof(PartialTypeCompletionProvider),
+                typeof(XmlDocCommentCompletionProvider),
+                typeof(TupleNameCompletionProvider),
+                typeof(DeclarationNameCompletionProvider),
+                typeof(InternalsVisibleToCompletionProvider),
+                typeof(PropertySubpatternCompletionProvider),
+                typeof(TypeImportCompletionProvider),
+                typeof(ExtensionMethodImportCompletionProvider),
+
+                // Built-in interactive providers
+                typeof(LoadDirectiveCompletionProvider),
+                typeof(ReferenceDirectiveCompletionProvider),
+                typeof(CSharpReplCommandCompletionProvider),
+
+                // Marker for end of built-in completion providers
+                typeof(LastBuiltInCompletionProvider),
+            };
+
+            AssertEx.EqualOrDiff(
+                string.Join(Environment.NewLine, expectedOrder.Select(x => x.FullName)),
+                string.Join(Environment.NewLine, actualOrder.Select(x => x.FullName)));
+        }
+
+        [Fact]
+        public void TestCompletionProviderOrderMetadata()
```

Comment:
so we test the strict ordering above?

Reply:
We test the complete order above, but the sorting logic is based on a partial order. This method tests for deterministic order.

Reply:
Added comments to the tests